### PR TITLE
cdc: flush meta correctly when flush log

### LINKED ISSUE #6189: ticdc redo apply shouldn't depend on meta
### Before asking a question, make sure you have

- [X] Searched existing Stack Overflow questions
- [X] Googled your question
- [X] Searched open and closed [GitHub issues](https://github.com/pingcap/tiflow/issues?q=is%3Aissue)
- [X] Read the [documentation](https://docs.pingcap.com/tidb/stable)

### What is your question?

In the current implement, when applying redo log on a given downstream sink, the event timestamp range is specified from `(meta.checkpointTs, meta.resolvedTs]`.

redo-log meta file is flushed periodically. However, events can be flushed into downstream sinks after **log file** is flushed instead of **redo-log meta**. It makes `cdc redo apply` not be able to ensure downstream integration. For example:

1. cdc table redo log is flushed with timestamp `ts1`;
2. however the current redo-meta resolved timestamp is `ts0`, which is less than `ts1`;
3. events before `ts1` have been flushed into downstream sinks;
4. cdc is killed and `cdc redo apply` is called with the current redo-meta.
5. after recovered, the downstream can contain events after the point `ts0`, which breaks the required integration.

I have prepared a [branch](https://github.com/hicqu/ticdc/tree/6.1.0-for-oncall-5117), and please notice the [fatal log](https://github.com/hicqu/ticdc/blob/6.1.0-for-oncall-5117/cdc/sink/mysql/mysql.go#L236). And the panic really happens when I run the branch:
```
[2022/07/05 16:00:27.117 +08:00] [FATAL] [mysql.go:237] ["[QP] mysqlSink.FlushRowChangedEvents is called with ts less than LogWriterResolved"] [stack="github.com/pingcap/tiflow/cdc/sink/mysql.(*mysqlSink).FlushRowChangedEvents\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql/mysql.go:237\ngithub.com/pingcap/tiflow/cdc/sink.(*tableSink).flushResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/sink/table_sink.go:109\ngithub.com/pingcap/tiflow/cdc/sink.(*tableSink).FlushRowChangedEvents\n\tgithub.com/pingcap/tiflow/cdc/sink/table_sink.go:86\ngithub.com/pingcap/tiflow/cdc/processor/pipeline.(*sinkNode).flushSink\n\tgithub.com/pingcap/tiflow/cdc/processor/pipeline/sink.go:155\ngithub.com/pingcap/tiflow/cdc/processor/pipeline.(*sinkNode).updateBarrierTs\n\tgithub.com/pingcap/tiflow/cdc/processor/pipeline/sink.go:338\ngithub.com/pingcap/tiflow/cdc/processor/pipeline.(*sinkNode).HandleMessage\n\tgithub.com/pingcap/tiflow/cdc/processor/pipeline/sink.go:329\ngithub.com/pingcap/tiflow/cdc/processor/pipeline.(*sinkNode).Receive\n\tgithub.com/pingcap/tiflow/cdc/processor/pipeline/sink.go:289\ngithub.com/pingcap/tiflow/pkg/pipeline.(*nodeRunner).run\n\tgithub.com/pingcap/tiflow/pkg/pipeline/runner.go:65\ngithub.com/pingcap/tiflow/pkg/pipeline.(*Pipeline).driveRunner\n\tgithub.com/pingcap/tiflow/pkg/pipeline/pipeline.go:97"]
``` 

The test shows that the above guess is correct.

To resolve the critical bug, we can consider to 
* always flush log meta when flushing log files, or
* remove log meta from the design, and put `checkpointTs` and `resolvedTs` into redo-log entry

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

Issue Number: close #6189 

### What is changed and how it works?

When flushing redo log, also flush meta with correct resolved timestamp.

#### Tests 

 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/redo/writer/writer.go [hunk 1]
```diff
@@ -549,8 +550,8 @@ func (l *LogWriter) setMaxCommitTs(tableID int64, commitTs uint64) uint64 {
 }
 
 // flush flushes all the buffered data to the disk.
-func (l *LogWriter) flush() error {
-	err1 := l.flushLogMeta(0, 0)
+func (l *LogWriter) flush(minResolvedTs uint64) error {
+	err1 := l.flushLogMeta(0, minResolvedTs)
```
Comment (CharlesCheung96) [inline cdc/redo/writer/writer.go:554]:
```suggestion
	err1 := l.ddlWriter.Flush()
	err2 := l.rowWriter.Flush()
	err3 := l.flushLogMeta(0, minResolvedTs)
```
Data should be flushed first.

File: cdc/redo/manager.go [hunk 2]
```diff
@@ -391,15 +391,30 @@ func (m *ManagerImpl) flushLog(
 	go func() {
 		defer atomic.StoreInt64(&m.flushing, 0)
 
-		err := m.writer.FlushLog(ctx, tableRtsMap)
+		var minResolvedTs uint64 = 0
+		if len(tableRtsMap) > 0 {
+			minResolvedTs = math.MaxUint64
+			for _, resolvedTs := range tableRtsMap {
```
Comment (CharlesCheung96) [inline cdc/redo/manager.go:397]: Consider the following case:
1. current state: `m.rtsMap` manage two tables: t1, t2. And `m.rtsMap[t1] = 1, m.rtsMap[t2] = 2`, `m.minResolvedTs = 1`.
2. `tableRtsMap = {"t2": 3}`
 Then, `minResolvedTs` computed here will be 3， which is too large.

Maybe, we should calculate `minResolvedTs` in `bgUpdateLog`.

File: cdc/redo/manager.go [hunk 3]
```diff
@@ -310,9 +342,10 @@ func (m *ManagerImpl) EmitDDLEvent(ctx context.Context, ddl *model.DDLEvent) err
 
 // GetResolvedTs returns the resolved ts of a table
 func (m *ManagerImpl) GetResolvedTs(tableID model.TableID) model.Ts {
-	m.rtsMapMu.Lock()
-	defer m.rtsMapMu.Unlock()
-	return m.rtsMap[tableID]
+	if value, ok := m.rtsMap.Load(tableID); ok {
+		return value.(*statefulRts).getFlushed()
+	}
+	return 0
```
Comment (CharlesCheung96) [inline cdc/redo/manager.go:348]: Normally it doesn't happen. Maybe we should print an error message here.

File: cdc/redo/manager.go [hunk 4]
```diff
@@ -452,8 +497,7 @@ func (m *ManagerImpl) bgUpdateLog(ctx context.Context, errCh chan<- error) {
 		case <-ticker.C:
 			// interpolate tick message to flush writer if needed
 			// TODO: add log and metrics
```
Comment (CharlesCheung96) [inline cdc/redo/manager.go:499]:
```suggestion
```

File: cdc/redo/manager.go [hunk 5]
```diff
@@ -384,45 +446,29 @@ func (m *ManagerImpl) flushLog(
...
+		m.postFlush(tableRtsMap, minResolvedTs)
 	}()
 
-	emptyRtsMap := make(map[model.TableID]model.Ts)
-	return emptyRtsMap
+	return
+}
+
+func (m *ManagerImpl) onResolvedTsMsg(tableID model.TableID, resolvedTs model.Ts) {
+	value, _ := m.rtsMap.LoadOrStore(tableID, &statefulRts{flushed: 0, unflushed: 0})
```
Comment (CharlesCheung96) [inline cdc/redo/manager.go:470]: Better to check `ok` here and panic if `ok == false`

File: cdc/redo/manager.go [hunk 6]
```diff
@@ -371,11 +398,46 @@ func (m *ManagerImpl) Cleanup(ctx context.Context) error {
...
-func (m *ManagerImpl) flushLog(
-	ctx context.Context,
-	tableRtsMap map[model.TableID]model.Ts,
-	handleErr func(err error),
-) map[model.TableID]model.Ts {
+func (m *ManagerImpl) prepareForFlush() (tableRtsMap map[model.TableID]model.Ts, minResolvedTs model.Ts) {
+	tableRtsMap = make(map[model.TableID]model.Ts)
+	minResolvedTs = math.MaxUint64
+	var minFlushedTs uint64 = math.MaxUint64
+	m.rtsMap.Range(func(key interface{}, value interface{}) bool {
```
Comment (CharlesCheung96) [inline cdc/redo/manager.go:405]:
```suggestion
	m.rtsMap.Range(func(key interface{}, value interface{}) bool {
		tableID := key.(model.TableID)
		rts := value.(*statefulRts)
		unflushed := rts.getUnflushed()
		flushed := rts.getFlushed()
		if unflushed > flushed {
			tableRtsMap[tableID] = unflushed
			flushed = unflushed
		}  
		
		if flushed < minResolvedTs {
			minResolvedTs = flushed
		}
		return true
	})

	if minResolvedTs == math.MaxUint64 {
		minResolvedTs = 0
	}
```

**Review by zhaoxinyu — APPROVED** [review]

**Pull request merged**